core/src/main/scala/filodb.core/downsample/DownsampledTimeSeriesShard.scala

```diff
@@ -313,37 +314,41 @@ class DownsampledTimeSeriesShard(rawDatasetRef: DatasetRef,
   }
 
   /**
-    * Iterator for lazy traversal of partIdIterator, value for the given label will be extracted from the ParitionKey.
+    * Iterator for traversal of partIds, value for the given label will be extracted from the ParitionKey.
+    * this implementation maps partIds to label/values eagerly, this is done inorder to dedup the results.
     */
   case class LabelValueResultIterator(partIds: debox.Buffer[Int], labelNames: Seq[String], limit: Int)
     extends Iterator[Map[ZeroCopyUTF8String, ZeroCopyUTF8String]] {
-    var currVal: Map[ZeroCopyUTF8String, ZeroCopyUTF8String] = _
-    var numResultsReturned = 0
-    var partIndex = 0
+    var iterIndx = 0
+    private lazy val rows = labelValues
 
-    override def hasNext: Boolean = {
-      var foundValue = false
-      while(partIndex < partIds.length && numResultsReturned < limit && !foundValue) {
-        val partId = partIds(partIndex)
+    def labelValues: Seq[Map[ZeroCopyUTF8String, ZeroCopyUTF8String]] = {
+      var partLoopIndx = 0
+      val rows = new mutable.HashSet[Map[ZeroCopyUTF8String, ZeroCopyUTF8String]]()
```

Comment:
Can you explain why we are using Set of Map here ?
Wont simple arraylist of values for each label work better here?

Reply:
since each row is a collection of key/values, we need Map here